clients/include/utility.hpp

```diff
@@ -2554,6 +2554,270 @@ void host_csrmm(J                     M,
+        {
+            host_csrmm(M,
+                       N,
+                       K,
+                       transA,
+                       transB,
+                       alpha,
+                       csr_row_ptr_A,
+                       csr_col_ind_A,
+                       csr_val_A,
+                       B + batch_stride_B * i,
+                       ldb,
+                       beta,
+                       C + batch_stride_C * i,
+                       ldc,
+                       order,
+                       base);
+        }
+    }
+    else if(Ci_Ai_B)
+    {
+        for(J i = 0; i < batch_count_C; i++)
+        {
+            host_csrmm(M,
+                       N,
+                       K,
+                       transA,
+                       transB,
+                       alpha,
+                       csr_row_ptr_A + offsets_batch_stride_A * i,
+                       csr_col_ind_A + columns_values_batch_stride_A * i,
+                       csr_val_A + columns_values_batch_stride_A * i,
+                       B,
+                       ldb,
+                       beta,
+                       C + batch_stride_C * i,
+                       ldc,
+                       order,
+                       base);
+        }
+    }
+    else if(Ci_Ai_Bi)
+    {
+        for(J i = 0; i < batch_count_C; i++)
+        {
+            host_csrmm(M,
+                       N,
+                       K,
+                       transA,
+                       transB,
+                       alpha,
+                       csr_row_ptr_A + offsets_batch_stride_A * i,
+                       csr_col_ind_A + columns_values_batch_stride_A * i,
+                       csr_val_A + columns_values_batch_stride_A * i,
+                       B + batch_stride_B * i,
+                       ldb,
+                       beta,
+                       C + batch_stride_C * i,
+                       ldc,
+                       order,
+                       base);
+        }
+    }
+}
+
+template <typename T, typename I>
+void host_coomm(I                    M,
+                I                    N,
+                I                    nnz,
+                hipsparseOperation_t transB,
+                T                    alpha,
+                const I*             coo_row_ind_A,
+                const I*             coo_col_ind_A,
+                const T*             coo_val_A,
+                const T*             B,
+                I                    ldb,
+                T                    beta,
+                T*                   C,
+                I                    ldc,
+                hipsparseOrder_t     order,
+                hipsparseIndexBase_t base)
+{
+    for(I j = 0; j < N; j++)
+    {
+#ifdef _OPENMP
+#pragma omp parallel for schedule(dynamic, 1024)
+#endif
+        for(I i = 0; i < M; ++i)
+        {
+            I idx_C  = order == HIPSPARSE_ORDER_COLUMN ? i + j * ldc : i * ldc + j;
```

Comment:
would be better to put this out of the loop and having 2 loops. In one case we have an awful traversal.

Reply:
FIxed